Meta/run.py

```diff
@@ -0,0 +1,978 @@
+from enum import Enum, unique
+from typing import cast
+import re
+
+QEMU_MINIMUM_REQUIRED_MAJOR_VERSION = 6
+QEMU_MINIMUM_REQUIRED_MINOR_VERSION = 2
+
+
+class RunError(Exception):
+    pass
+
+
+@unique
+class Arch(Enum):
+    """SerenityOS architecture, not host architecture."""
+
+    Aarch64 = "aarch64"
+    x86_64 = "x86_64"
+
+
+@unique
+class QEMUKind(Enum):
+    """VM distrinctions determining which virtualizers *might* be used."""
+
+    # Linux and anything that may or may not have KVM, including WSL with Linux QEMU.
+    Other = "-accel kvm"
+    # WSL with native Windows QEMU (and possibly WHPX).
+    NativeWindows = "-accel whpx"
+    # MacOS with HVF.
+    MacOS = "-accel hvf"
+    # Serenity on Serenity with ported QEMU
+    SerenityOS = "-accel tcg"
+
+
+@unique
+class MachineType(Enum):
+    Default = ""
+    QEMUTap = "qtab"
+    QEMUWithoutNetwork = "qn"
+    QEMUExtLinux = "qextlinux"
+    QEMU35 = "q35"
+    QEMU35Grub = "q35grub"
+    QEMUGrub = "qgrub"
+    CI = "ci"
+    Limine = "limine"
+    Bochs = "b"
+    MicroVM = "microvm"
+    ISAPC = "isapc"
+
+    def uses_grub(self) -> bool:
+        return self in [MachineType.QEMU35Grub, MachineType.QEMUGrub]
+
+    def is_q35(self) -> bool:
+        return self in [MachineType.QEMU35Grub, MachineType.QEMU35]
+
+
+@dataclass
+class Configuration:
+    """Run configuration, populated from command-line or environment variable data."""
+
+    # Binary and environmental configuration
+    virtualization_support: bool = False
+    bochs_binary: str = "bochs"
+    qemu_binary: str | None = None
+    qemu_kind: QEMUKind | None = None
+    kvm_usable: bool | None = None
+    architecture: Arch | None = None
+    enable_gdb: bool = False
+    enable_gl: bool = False
+    nvme_enable: bool = True
+    sd_enable: bool = False
+
+    # QEMU arguments
+    virtualization_args: str | None = None
+    kernel_cmdline: list[str] = field(default_factory=lambda: ["hello"])
+    ram_size: str = "1G"
+    disk_image: str = "_disk_image"
+    qemu_cpu: str = "max"
+    cpu_count: int = 2
+    extra_arguments: list[str] = field(default_factory=list)
+    machine_type: MachineType = MachineType.Default
+    spice_arguments: list[str] = field(default_factory=list)
+    audio_backend: str = "pa,timer-period=2000"
+    audio_device: list[str] = field(default_factory=list)
+    screen_count: int = 1
+    display_backend: str | None = None
+    display_device: str = "VGA,vgamem_mb=64"
+    boot_drive: list[str] = field(default_factory=list)
+    host_ip: str = "127.0.0.1"
+    ethernet_device_type: str = "e1000"
+    network_backend: str | None = None
+    network_default_device: list[str] = field(default_factory=list)
+    kernel_and_initrd: list[str] = field(default_factory=list)
+
+
+def kvm_usable() -> bool:
+    return access("/dev/kvm", os.R_OK | os.W_OK | os.X_OK)
+
+
+def determine_qemu_kind() -> QEMUKind:
+    if run(["command", "-v", "wslpath"], shell=True).returncode == 0:
```

Comment:
I would go for `shutil.which("wslpath")` here.

Reply:
Had that import around for a while but forgot to change it...